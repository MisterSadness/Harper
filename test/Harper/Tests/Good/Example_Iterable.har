value List a = {
  variant Empty = {
    isEmpty :: Bool;
    isEmpty = true;

    // Implementing iterate is a special case as compared to normal functions.
    // If we're to treat List a as an Iterable a, all variants must declare
    // an iterate function.
    iterate :: Iterator a;
    iterate = { yield return; };
  };
  
  variant NonEmpty = {
    data = {
      head :: a;
      tail :: List a;
    }
    
    isEmpty :: Bool;
    isEmpty = false;

    iterate :: Iterator a;
    iterate = {
      // The `this` identifier binds to the implicit first argument of a member function.
      var (current :: List a) = this;
      while not current.isEmpty {
        yield current.head;
        current := current.tail;
      }
    };
  };
};

cons :: a -> List a -> List a;
cons a l = val NonEmpty { head = a, tail = l };

// This can be generalised to Iterable a -> ().
printList :: List a -> sideeffect -> ();
printList xs = {
  for (x :: a) in xs {
    eval print x ();
    eval print " " ();
  }
  eval print "\n" ();
};

filter :: (a -> Bool) -> Iterable a -> Iterable a;
filter f xs = {
  for (x :: a) in xs {
    if f x {
	    yield x;
	  }
  }
};

map :: (a -> b) -> Iterable a -> Iterable b;
map f xs = {
  for (x :: a) in xs {
    yield f x;
  }
};

foldl :: (a -> b -> b) -> b -> Iterable a -> b;
foldl f acc xs = {
  var (result :: b) = acc;
  for (x :: a) in xs {
    result := f x result;
  }
  return result;
};

main :: sideeffect -> ();
main = {
  (sum :: Integer -> Integer -> Integer) = \(x :: Integer) (y :: Integer) => x + y;
  (sqr :: Integer -> Integer) = \(x :: Integer) => x ^ 2;
  (pred :: Integer -> Bool) = \(x :: Integer) => x != 0;
  eval printLn (foldl sum 0 | map sqr | filter pred | cons 42 | cons 17 | cons 37 | cons 0 (val Empty { })) ();
};