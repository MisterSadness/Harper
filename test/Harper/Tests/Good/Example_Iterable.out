3422

Execution ended with value:
()